### PR TITLE
Prep Page for removal of InternalsVisibleTo 

### PR DESCRIPTION
**Pull request by hartez**
### Description of Change ###

Add IPageController interface and update renderers to use it

### CODE REVIEW

File: Xamarin.Forms.Core/Element.cs [hunk 1]
```diff
@@ -107,10 +107,9 @@ public string StyleId
 			}
 		}
 
-		internal virtual ReadOnlyCollection<Element> LogicalChildren
-		{
-			get { return EmptyChildren; }
-		}
+		internal virtual ReadOnlyCollection<Element> LogicalChildrenInternal => EmptyChildren;
```
Comment (kingces95) [inline Xamarin.Forms.Core/Element.cs]: You may have already considered this but interface definitions can be "overridden" like a virtual method by declaring the interface on the derived class and in that way we wouldn't need `Internal` methods. There are drawbacks such as having to implement all the other interface methods and not being as explicit without the `override` keyword. So up to you. https://gist.github.com/kingces95/47a4308e1597e149a214eb6e0daa9d5d

File: Xamarin.Forms.Core/MultiPage.cs [hunk 2]
```diff
@@ -199,22 +199,24 @@ void OnChildrenChanged(object sender, NotifyCollectionChangedEventArgs e)
 
 		void OnTemplatedItemsChanged(object sender, NotifyCollectionChangedEventArgs e)
 		{
+			var pageController = ((IPageController)this);
```
Comment (kingces95) [inline Xamarin.Forms.Core/MultiPage.cs]: Might consider expanding the idea an making a private property that preforms the cast: `IPageController Controller => this`

**Comment by kingces95** [comment]
:+1: with a few suggestions

File: Xamarin.Forms.Core/Element.cs (hunk 1 above)
Comment (hartez) [inline Xamarin.Forms.Core/Element.cs]: Seems like that would be problematic for anything that needs to use `base`. It'd be doable, but kind of a pain for maintenance.

File: Xamarin.Forms.Core/MultiPage.cs (hunk 2 above)
Comment (hartez) [inline Xamarin.Forms.Core/MultiPage.cs]: Yeah, that'd look a lot cleaner (and be easier to maintain). I'll do that.

Comment (kingces95) [inline Xamarin.Forms.Core/MultiPage.cs]: Cool, you gonna make that change across all your ivt checkins?

Comment (hartez) [inline Xamarin.Forms.Core/MultiPage.cs]: Yep.

File: Xamarin.Forms.Core/MultiPage.cs [hunk 3]
```diff
@@ -305,7 +308,7 @@ void OnTemplatedItemsChanged(object sender, NotifyCollectionChangedEventArgs e)
 
 		void Reset()
 		{
-			List<Element> snapshot = InternalChildren.ToList();
+			List <Element> snapshot = InternalChildren.ToList();
```
Comment (rmarinho) [inline Xamarin.Forms.Core/MultiPage.cs]: space here not relevant

File: Xamarin.Forms.Platform.Android/Renderers/MasterDetailRenderer.cs [hunk 4]
```diff
@@ -254,20 +258,14 @@ void HandlePropertyChanged(object sender, PropertyChangedEventArgs e)
 
 		void MasterDetailPageAppearing(object sender, EventArgs e)
 		{
-			if (_page.Master != null)
-				_page.Master.SendAppearing();
-
-			if (_page.Detail != null)
-				_page.Detail.SendAppearing();
+			MasterPageController?.SendAppearing();
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.Android/Renderers/MasterDetailRenderer.cs]: One has to love ?.  :)

**Comment by rmarinho** [comment]
👎 failing a UITest on android,

**Comment by samhouts** [comment]
:+1: 6 green lights!

**Pull request merged**